### PR TITLE
okhttp: local-only transparent retry for okhttp

### PR DESCRIPTION
**Pull request by dapengzhang0**
The OKHttp counterpart for #8878.

### CODE REVIEW

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpClientStream.java [hunk 1]
```diff
@@ -372,7 +373,8 @@ private void cancel(Status reason, boolean stopDelivery, Metadata trailers) {
         requestHeaders = null;
         pendingData.clear();
         canStart = false;
-        transportReportStatus(reason, true, trailers != null ? trailers : new Metadata());
+        transportReportStatus(
+            reason, MISCARRIED, true, trailers != null ? trailers : new Metadata());
```
Comment (ejona86) [inline okhttp/src/main/java/io/grpc/okhttp/OkHttpClientStream.java:377]: So this is a case where cancel() is called before start()? I don't think we need retries at all for that case. Seems best to keep this as-it-was.

**Comment by dapengzhang0** [comment]
> codecov/patch — 66.66% of diff hit (target 80.00%)

Seems only line 

https://github.com/grpc/grpc-java/pull/8894/files#diff-6f6e764b37aa97214fb12bc87a4c11b16e23262952bb0458aa62375f43f99a9cR422

is not covered by test, neither the original version. I could not find a way to easily produce this scenario unless over engineering it.

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java [hunk 2]
```diff
@@ -811,7 +811,8 @@ public void shutdownNow(Status reason) {
       }
 
       for (OkHttpClientStream stream : pendingStreams) {
-        stream.transportState().transportReportStatus(reason, true, new Metadata());
+        stream.transportState().transportReportStatus(
+            reason, RpcProgress.MISCARRIED, true, new Metadata());
```
Comment (ejona86) [inline okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java:815]: This is in shutdownNow(). That's similar to this stream being explicitly cancelled. I don't think we should change this location, and I'm not aware of any need for it.

Comment (dapengzhang0) [reply okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java:815]: If a connection fails to ack keep-alive, the transport will `shutdownNow`. I think the pending streams should have a chance to be routed to another connection.

**Comment by ejona86** [comment]
> Seems only line
> 
> https://github.com/grpc/grpc-java/pull/8894/files#diff-6f6e764b37aa97214fb12bc87a4c11b16e23262952bb0458aa62375f43f99a9cR422
> 
> is not covered by test, neither the original version. I could not find a way to easily produce this scenario unless over engineering it.

That looks like the case where the transport has received GOAWAY and a new stream is created (a race with transportShutdown, when integrated). Dunno how hard that would be to make a test for.

File: okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java (hunk 2 above)
Comment (ejona86) [reply okhttp/src/main/java/io/grpc/okhttp/OkHttpClientTransport.java:815]: I see. Sounds good.

**Review by ejona86 — APPROVED** [review]

**Comment by dapengzhang0** [comment]
I'll hold on merging it until #8878 is imported internally and not causing any issues.

**Pull request merged**